Board

rlcheng opened the pull request:
Got a lot of the board logic done. Can set marker and it can dynamically create lists of winning combinations for rows, columns and diagonals. Going to use that for the determine-winner method so that's still in progress while I read up Clojure for the Brave and True.

Commit:
Refactor existing code for clarity. Add ability to create new board based on width. Add determine-winner function. Incomplete but resulted in creation of winning-rows function that lists all possible win row combinations.

rlcheng committed:
Add winning diagonal and minor style update in board_spec

dcmoore commented on src/tic_tac_toe/board.clj:
```diff
@@ -2,10 +2,30 @@
+	(* width width))
+
+(defn get-new-board [width]
+  vector (replicate (get-size width) space))
```
Does this work? Looks like you're missing a paren around the vector function.

rlcheng replied:
Just checked. if I put paren around the vector function and run the tests i get the following:
```
Failures:

  1) create empty board should create a board with size of 9 if width is 3
     Expected: [" " " " " " " " " " " " " " " " " "]
          got: [(" " " " " " " " " " " " " " " " " ")] (using =)
     /Users/richardcheng/Documents/clojure-tic-tac-toe/spec/tic_tac_toe/board_spec.clj:20

  2) create empty board should create a board with size of 16 if width is 4
     Expected: [" " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " "]
          got: [(" " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " ")] (using =)
     /Users/richardcheng/Documents/clojure-tic-tac-toe/spec/tic_tac_toe/board_spec.clj:22

Finished in 0.00970 seconds
13 examples, 2 failures
```

dcmoore replied:
What's the point of the vector function if you're not going to call it?

rlcheng replied:
Updated. Tested some more and realized i had a misconception on that output of replicate. also changed it to repeat.

rlcheng committed:
Remove vector from get-new-board because it is not being used. Change replicate to repeat because replicate is obsolete according to cllojure docs.

dcmoore commented on .nrepl-port:
```diff
@@ -0,0 +1 @@
+50301
```
Why is this file committed? Isn't this user specific?

rlcheng replied:
commited by accident

dcmoore commented on src/tic_tac_toe/board.clj:
```diff
@@ -2,10 +2,30 @@
 
```
What is the single responsibility of the board namespace? Do you adhere to SRP as it stands?

rlcheng committed:
update gitignore and remove .nrepl-port

rlcheng committed:
Add full? function to determine if board is full. Returns true if full, otherwise returns false.

rlcheng committed:
Refactor win determination logic to score_keeper and added methods and logic to make winner? method fully functional.
- Add winning-combos method that outputs all win combinations of given board width.
- Add get-positions method that retrieves markers of current board given list of positions.
- Add has-same-marker? method that checks if marker matches given collection of row, column or diagonal marker list.
- Rename and add logic to winner? that given current board, marker, and width if the marker has won.

tbaik commented on src/tic_tac_toe/board.clj:
```diff
@@ -2,10 +2,16 @@
     nil))
+
+(defn full? [board]
+	(not-any? #(= space %) board))
```
Small nitpick: It seems like some of these have 2 spaces and some have more (4 or 6?) This might bug some people with OCD 😄

tbaik commented on spec/tic_tac_toe/score_keeper_spec.clj:
```diff
@@ -0,0 +1,88 @@
+  (it "should return list of winning diagonal combinations if width is 3"
+    (should= '((0 4 8) (2 4 6)) (winning-diagonals 3)))
+  (it "should return list of winning diagonal combinations if width is 4"
+    (should= '((0 5 10 15) (3 6 9 12)) (winning-diagonals 4))))
```
For all the tests for winning rows, columns, and diagonals, I would recommend testing such behavior through testing the #winner? function.

Read this: https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html.

We want to push for testing the behavior of the program. More specifically, the external facing behavior of this module. In this case, #winning-rows, #winning-columns, and #winning-diagonals seem to only be internal facing (private functions opposed to public).

I would approach writing tests with each of these tests you wrote for these for the #winner? function instead, making it pass, and refactoring it eventually into these three private functions.

I initially struggled to understand why someone would suggest something like this previously. We can have a further chat in slack if you're feeling the same!

tbaik commented on src/tic_tac_toe/score_keeper.clj:
```diff
@@ -0,0 +1,29 @@
+	(for [x (range 0 width)]
+		(range x (board/get-size width) width)))
+
+(defn winning-diagonals [width]
```
You can use `(defn-` for all of your private functions in this namespace

Akcbryant commented on src/tic_tac_toe/score_keeper.clj:
```diff
@@ -0,0 +1,29 @@
+
+(defn winning-cols [width]
+  (for [x (range 0 width)]
+    (range x (board/get-size width) width)))
```
I wonder what this might look like using a map instead of a for loop? I generally think `map` is more idiomatic to Clojure.

Akcbryant commented on src/tic_tac_toe/score_keeper.clj:
```diff
@@ -0,0 +1,29 @@
+    (for [x (range 0 (board/get-size width) (inc width))]
+      x)
+    (for [x (range 1 (inc width))]
+      (* x (dec width)))))
```
Same as above.

Akcbryant commented on src/tic_tac_toe/score_keeper.clj:
```diff
@@ -0,0 +1,29 @@
+(defn get-positions [board positions]
+  (for [x positions]
+    (get board x)))
+
```
Here as well.

Akcbryant commented on spec/tic_tac_toe/score_keeper_spec.clj:
```diff
@@ -0,0 +1,88 @@
+    (should= true (has-same-marker? X-row-3 X))
+    (should= false (has-same-marker? mix-row-3 X))
+    (should= true (has-same-marker? O-row-4 O))
+    (should= false (has-same-marker? mix-row-4 O))))
```
I generally try to stick to one `it` statement for each assertion. This makes it easier to pinpoint exactly where our problem is.

Akcbryant commented on spec/tic_tac_toe/score_keeper_spec.clj:
```diff
@@ -0,0 +1,88 @@
+(def O-fourth-col-win-4x4 [" " " " " " O " " " " " " O " " " " " " O " " " " " " O])
+(def X-first-diag-win-4x4 [X " " " " " " " " X " " " " " " " " X " " " " " " " " X])
+(def O-second-diag-win-4x4 [" " " " " " O " " " " O " " " " O " " " " O " " " " " "])
+(def stalemate-3x3 [X O X O O X X X O])
```
It might help to put the board arrays in a more readable format:
```
[ X O X
  O O X
  X X O ]
```

Akcbryant replied:
In this format it might be less confusing to have these inline with your tests.

Akcbryant commented on spec/tic_tac_toe/score_keeper_spec.clj:
```diff
@@ -0,0 +1,88 @@
+    (should= true (winner? O-second-diag-win-4x4 O 4))
+    (should= nil (winner? empty-board-3x3 X 3))
+    (should= nil (winner? stalemate-3x3 X 3))
+    (should= nil (winner? stalemate-3x3 O 3))))
```
This list is going to just keep growing with each new board size. How can this be more generic so we don't have to explicitly add tests cases for future board sizes?

Akcbryant reviewed (commented):
Let me know if you have any questions.

rlcheng committed:
Change board array format in tests to make them more readable.

rlcheng committed:
Refactor:
 - Refactor score_keeper to use map instead of for loops.
 - Clean up tests. Add more detailed tests while eliminating unnecessary ones.
 - winner? method. Remove width argument. Add in variables for readability.
 - Add get-width method to board to achieve removing width from winner? method.
 - winner? method returns false instead of nil.
 - Change method names to include verbs where needed.
 - Rename has-same-marker? to same-marker? so method names are uniform for all methods with question marks.
 - Rename score_keeper to board_evaluator.

rlcheng committed:
- Refactor/move #get-positions from board-evaluator to board since method is a board functionality.
- Refactor multiple methods in board-evaluator to be private methods, leaving only #winner? to be used externally. Refactor tests to reflect change.

tbaik commented:
👍

Akcbryant commented:
👍
